Add 1+, 1- and -1+, and improve `abs` in cptypes

gus-massa opened the pull request:
Two slightly relate patchs.

**Add 1+, 1- and -1+ to cptypes**

I should have added them in the (old) previous commit that added `add1` and `sub1`, but I forget these primitives existed. I refactored some of the test to avoid too much duplication.

**Improve fixnum case of abs in cptypes**

In the (old) previous commit I added a reduction for `(abs <flnumber>)` to `(flabs <flnumber>)`. But it has no reduction for `(abs <fxnumber>)`. 

Now I'm reducing `(abs <fx>)` to

    (let ([t <fx>])
      (if (fx= t (most-negative-fixnum))
          (pariah (- (most-negative-fixnum)))
          (fxabs t)))

that is the straightforward expansion. It's expanded later in `cpprim` to  

    (let ([t <fx>])
      (if (fx= t (most-negative-fixnum))
          (- (most-negative-fixnum))
          (if (fx>= t 0) t (fx- t))))

But perhaps it's better to expand it in `cptypes` directly to

    (let ([t <fx>])
      (if (fx>= t 0)
          t
          (if (fx= t (most-negative-fixnum))
              (- (most-negative-fixnum))
              (fx- t))))

becuse it is faster for positive numbers and slower for `(most-negative-fixnum)` that is very unusual.

I also wanted to add a similar special case for `add1` and `sub1`, but in their expansion in `cpprim` they use the `overflow` flag, so expanding them to a comparison to `(most-negative-fixnum)` or `(most-positive-fixnum)` is probably slower. (I hope to try something like this next month.)

burgerrg commented:
Thank you for the updates! Please add a subsection to the release notes describing them.

gus-massa commented:
I fixed one misplace parenthesis that was causing an error in one of the tests.

I added a note for each commit. It is it fine or it's better to accumulate all the changes in a single note? 

But I did't add this text in the notes to avoid repetition. I can add it if you think it's necessary.

> Type recovery is enabled with the \scheme{enable-type-recovery} parameter.

burgerrg commented on release_notes/release_notes.stex:
```diff
@@ -116,6 +116,14 @@ Online versions of both books can be found at
 %-----------------------------------------------------------------------------
 \section{Functionality Changes}\label{section:functionality}
 
+\subsection{Improved support abs in type recovery (10.2.0)}
+
+The type recovery pass has improved support for abs when the argument is a fixnum.
+
+\subsection{Add support for 1+, 1- and -1+ in type recovery (10.2.0)}
+
+The type recovery pass has support for 1+, 1- and -1+.
+
```
I'd combine these into a single entry, something like:
```
\subsection{Type recovery improvements (10.2.0)}

The type recovery pass has improved support for \scheme{abs} with a fixnum argument
and added support for \scheme{1+}, \scheme{1-}, and \scheme{-1+}.
```

burgerrg replied:
It's fine not to add the sentence about enabling type recovery.

gus-massa commented:
Done

burgerrg commented on s/cptypes.ss:
```diff
@@ -298,10 +298,38 @@ Notes:
+                 (cdr e*) (cdr r*))]
+          [(try-ref->prelex/not-assigned (car e*))
+           => (lambda (v)
+                (set-prelex-multiply-referenced! v #t) ; just in case it was sinlge referenced
```
sinlge => singly

burgerrg commented on s/cptypes.ss:
```diff
@@ -298,10 +298,38 @@ Notes:
+      ; In the results:
+      ;   before*, var* and e* may be shorter than the arguments.
+      ;   var* and e* have the same length.
+      ;   ref* has the same lenght than the arguments.
```
lenght than => length as

gus-massa replied:
I fixed both typos and also a "lenght" somewhere else.

mflatt commented on mats/cptypes.ms:
```diff
@@ -880,97 +893,66 @@
-         '(lambda (x) (when (bignum? x)
-                        (bignum? (sub1 x))))
-         '(lambda (x) (when (bignum? x)
-                        #t))))
```
Should this test have been kept? I don't think it's covered by the `test-closed1` generalization, but I might be missing something else.

burgerrg replied:
No, because when x = (+ (most-positive-fixnum) 1), it's incorrect.

gus-massa replied:
I undeleted the test. It's a to ensure I don't make an easy mistake in the future. I don't know why I deleted it.

It's not included in `test-closed1`. Moreover, I was thinking about writing `test-not-closed1` to reduce duplication, but I decided that it was too weird. But Perhaps I'll add it in the future if there are too many cases of almost-closed functions.
